domains/client/block-preprocessor/src/inherents.rs

```diff
@@ -69,25 +69,40 @@ where
     CBlock: BlockT,
     Block: BlockT,
 {
-    let header = consensus_client.header(consensus_block_hash)?.ok_or(
-        sp_blockchain::Error::MissingHeader(format!(
-            "No header found for {consensus_block_hash:?}"
-        )),
-    )?;
-
     let runtime_api = consensus_client.runtime_api();
+
     let runtime_id = runtime_api
         .runtime_id(consensus_block_hash, domain_id)?
         .ok_or(sp_blockchain::Error::Application(Box::from(format!(
             "No RuntimeId found for {domain_id:?}"
         ))))?;
 
-    Ok(header
-        .digest()
-        .logs
-        .iter()
-        .filter_map(|log| log.as_domain_runtime_upgrade())
-        .any(|upgraded_runtime_id| upgraded_runtime_id == runtime_id))
+    // The runtime_upgrades() API is only present in API versions 2 and later. On earlier versions,
+    // we need to call legacy code.
+    // TODO: remove version check before next network
+    let domains_api_version = runtime_api
+        .api_version::<dyn DomainsApi<CBlock, CBlock::Header>>(consensus_block_hash)?
+        // It is safe to return a default version of 1, since there will always be version 1.
+        .unwrap_or(1);
+
+    let is_upgraded = if domains_api_version >= 2 {
+        let runtime_upgrades = runtime_api.runtime_upgrades(consensus_block_hash)?;
+        runtime_upgrades.contains(&runtime_id)
+    } else {
+        let header = consensus_client.header(consensus_block_hash)?.ok_or(
+            sp_blockchain::Error::MissingHeader(format!(
+                "No header found for {consensus_block_hash:?}"
+            )),
+        )?;
+        header
+            .digest()
+            .logs
+            .iter()
+            .filter_map(|log| log.as_domain_runtime_upgrade())
+            .any(|upgraded_runtime_id| upgraded_runtime_id == runtime_id)
+    };
```

Comment:
I think we'll need this compatibility code for inherents, so that runtime upgrades on Taurus will work. I avoided repeating the compatibility code by calling this method instead.

I think we'll be able to remove this code after the runtime on Taurus is upgraded, and most nodes on Taurus and Mainnet are upgraded?

Reply:
I didn't read into this PR to understand the context, but as a general rule inherents can't be removed without making it impossible to import older blocks. They are kind of like host functions for runtime environment that once added can't be removed.

Reply:
> I didn't read into this PR to understand the context, but as a general rule inherents can't be removed without making it impossible to import older blocks. They are kind of like host functions for runtime environment that once added can't be removed.

No inherents are added or removed in this PR, it just changes how we process them to avoid some duplicate work.

This PR changes the code the domain client calls when it processes a domain runtime upgrade inherent. Previously, we searched through the block header to determine if the runtime was upgraded. Now we ask the runtime for a list of upgraded runtime IDs (which the domains runtime has already determined by doing a similar search.)

Reply:
>I think we'll be able to remove this code (edit: the compatibility code) after the runtime on Taurus is upgraded, and most nodes on Taurus and Mainnet are upgraded.

Nazar is correct, we can't remove the compatibility code even after runtime/node upgraded. Because the runtime code is on-chain state, it applies to block that after the upgrade and before the next upgrade, older block will still use the old runtime (which doesn't have the new runtime API). Image a node (running on the latest release that doesn't contain the compatibility code) just joined the network and sync from genesis, it will fail to call the new API until the block that upgrade the runtime.